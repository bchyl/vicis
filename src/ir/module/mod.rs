pub mod attributes;
pub mod global_variable;
pub mod linkage;
pub mod name;
pub mod parser;
pub mod preemption_specifier;

pub use parser::parse as parse_assembly;

use super::{
    function::{Function, FunctionId},
    types::Types,
};
use attributes::Attribute;
use global_variable::GlobalVariable;
use id_arena::Arena;
use name::Name;
use rustc_hash::FxHashMap;
use std::fmt;

#[derive(Debug, Clone)]
struct Target {
    triple: String,
    datalayout: String,
}

pub struct Module {
    name: String,
    source_filename: String,
    target: Target,
    functions: Arena<Function>,
    attributes: FxHashMap<u32, Vec<Attribute>>,
    global_variables: FxHashMap<Name, GlobalVariable>,
    pub types: Types,
    // TODO: Metadata
}

impl Module {
    pub fn new() -> Self {
        Self {
            name: "".to_string(),
            source_filename: "".to_string(),
            target: Target::new(),
            functions: Arena::new(),
            attributes: FxHashMap::default(),
            global_variables: FxHashMap::default(),
            types: Types::new(),
        }
    }

    pub fn name(&self) -> &String {
        &self.name
    }

    pub fn functions(&self) -> &Arena<Function> {
        &self.functions
    }

    pub fn functions_mut(&mut self) -> &mut Arena<Function> {
        &mut self.functions
    }

    pub fn find_function_by_name<T: AsRef<str>>(&self, name: T) -> Option<FunctionId> {
        for (id, func) in &self.functions {
            if func.name() == name.as_ref() {
                return Some(id);
            }
        }
        None
    }
}

impl Target {
    pub fn new() -> Self {
        Self {
            triple: "".to_string(),
            datalayout: "".to_string(),
        }
    }
}

impl fmt::Debug for Module {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "source_filename = \"{}\"", self.source_filename)?;
        writeln!(f, "target datalayout = \"{}\"", self.target.datalayout)?;
        writeln!(f, "target triple = \"{}\"", self.target.triple)?;
        writeln!(f)?;
        for (_, gv) in &self.global_variables {
            writeln!(f, "{}", gv.to_string(&self.types))?;
        }
        writeln!(f)?;
        for (_, func) in &self.functions {
            writeln!(f, "{:?}", func)?;
        }
        for (id, attrs) in &self.attributes {
            write!(f, "attributes #{} = {{ ", id)?;
            for attr in attrs {
                write!(f, "{:?} ", attr)?;
            }
            writeln!(f, "}}")?
        }
        Ok(())
    }
}
