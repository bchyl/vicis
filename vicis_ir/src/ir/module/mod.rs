pub mod attributes;
pub mod global_variable;
pub mod linkage;
pub mod name;
pub mod parser;
pub mod preemption_specifier;
pub mod unnamed_addr;
pub mod visibility;

pub use parser::parse as parse_assembly;

use super::{
    function::{Function, FunctionId},
    types::Types,
};
use attributes::Attribute;
use global_variable::GlobalVariable;
use id_arena::Arena;
use name::Name;
use rustc_hash::FxHashMap;
use std::fmt;

#[derive(Debug, Clone)]
pub struct Target {
    triple: String,
    datalayout: String,
}

pub struct Module {
    pub(crate) name: String,
    pub(crate) source_filename: String,
    pub(crate) target: Target,
    pub(crate) functions: Arena<Function>,
    pub(crate) attributes: FxHashMap<u32, Vec<Attribute>>,
    pub(crate) global_variables: FxHashMap<Name, GlobalVariable>,
    pub types: Types,
    // TODO: Metadata
}

impl Default for Module {
    fn default() -> Self {
        Self {
            name: "".to_string(),
            source_filename: "".to_string(),
            target: Target::new(),
            functions: Arena::new(),
            attributes: FxHashMap::default(),
            global_variables: FxHashMap::default(),
            types: Types::new(),
        }
    }
}

impl Module {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn name(&self) -> &String {
        &self.name
    }

    pub fn functions(&self) -> &Arena<Function> {
        &self.functions
    }

    pub fn functions_mut(&mut self) -> &mut Arena<Function> {
        &mut self.functions
    }

    pub fn find_function_by_name<T: AsRef<str>>(&self, name: T) -> Option<FunctionId> {
        for (id, func) in &self.functions {
            if func.name() == name.as_ref() {
                return Some(id);
            }
        }
        None
    }
}

impl Default for Target {
    fn default() -> Self {
        Self {
            triple: "".to_string(),
            datalayout: "".to_string(),
        }
    }
}

impl Target {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn triple(&self) -> &str {
        self.triple.as_str()
    }

    pub fn datalayout(&self) -> &str {
        self.datalayout.as_str()
    }
}

impl fmt::Debug for Module {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "source_filename = \"{}\"", self.source_filename)?;
        writeln!(f, "target datalayout = \"{}\"", self.target.datalayout)?;
        writeln!(f, "target triple = \"{}\"", self.target.triple)?;
        writeln!(f)?;
        write!(f, "{:?}", self.types)?;
        for gv in self.global_variables.values() {
            writeln!(f, "{}", gv.to_string(&self.types))?;
        }
        writeln!(f)?;
        for (_, func) in &self.functions {
            writeln!(f, "{:?}", func)?;
        }
        for (id, attrs) in &self.attributes {
            write!(f, "attributes #{} = {{ ", id)?;
            for attr in attrs {
                write!(f, "{:?} ", attr)?;
            }
            writeln!(f, "}}")?
        }
        Ok(())
    }
}
