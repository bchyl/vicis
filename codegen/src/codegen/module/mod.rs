use super::{function::Function, isa::TargetIsa};
use id_arena::Arena;
use rustc_hash::FxHashMap;
use std::{fmt, marker::PhantomData};
use vicis_core::ir::{
    module::{
        attributes::Attribute, global_variable::GlobalVariable, name::Name, Module as IrModule,
    },
    types::Types,
};

pub struct Module<'a, T: TargetIsa> {
    pub module: &'a IrModule,
    pub functions: Arena<Function<'a, T>>,
    pub attributes: FxHashMap<u32, Vec<Attribute>>,
    pub global_variables: FxHashMap<Name, GlobalVariable>,
    pub types: Types,
    pub _isa: PhantomData<fn() -> T>,
}

impl<T: TargetIsa> fmt::Debug for Module<'_, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "source_filename = \"{}\"", self.module.source_filename())?;
        writeln!(
            f,
            "target datalayout = \"{}\"",
            self.module.target().datalayout()
        )?;
        writeln!(f, "target triple = \"{}\"", self.module.target().triple())?;
        writeln!(f)?;
        for gv in self.global_variables.values() {
            writeln!(f, "{}", gv.to_string(&self.types))?;
        }
        writeln!(f)?;
        for (_, func) in &self.functions {
            writeln!(f, "{:?}", func)?;
        }
        for (id, attrs) in &self.attributes {
            write!(f, "attributes #{} = {{ ", id)?;
            for attr in attrs {
                write!(f, "{:?} ", attr)?;
            }
            writeln!(f, "}}")?
        }
        Ok(())
    }
}
