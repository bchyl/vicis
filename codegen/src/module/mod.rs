use super::{function::Function, isa::TargetIsa};
use id_arena::Arena;
use std::fmt;
use vicis_core::ir::{module::Module as IrModule, types::Types};

pub struct Module<'a, T: TargetIsa> {
    pub ir: &'a IrModule,
    pub functions: Arena<Function<'a, T>>,
    pub types: Types,
    pub isa: &'a T,
}

impl<T: TargetIsa> fmt::Debug for Module<'_, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "source_filename = \"{}\"", self.ir.source_filename())?;
        writeln!(
            f,
            "target datalayout = \"{}\"",
            self.ir.target().datalayout()
        )?;
        writeln!(f, "target triple = \"{}\"", self.ir.target().triple())?;
        writeln!(f)?;
        for gv in self.ir.global_variables().values() {
            writeln!(f, "{}", gv.to_string(&self.types))?;
        }
        writeln!(f)?;
        for (_, func) in &self.functions {
            writeln!(f, "{:?}", func)?;
        }
        for (id, attrs) in self.ir.attributes() {
            write!(f, "attributes #{} = {{ ", id)?;
            for attr in attrs {
                write!(f, "{:?} ", attr)?;
            }
            writeln!(f, "}}")?
        }
        Ok(())
    }
}
