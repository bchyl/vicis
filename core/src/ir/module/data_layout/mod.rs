use crate::ir::types::{self, CompoundType, StructType, Type, Types};

#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
pub struct DataLayout(pub String);

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StructLayout(Vec<usize>, usize, usize);

impl DataLayout {
    pub fn new_struct_layout_for(&self, types: &Types, ty: Type) -> Option<StructLayout> {
        if let CompoundType::Struct(ref sty) = &*types.get(ty)? {
            return Some(StructLayout::new(self, types, sty));
        }
        None
    }

    /// Returns the size of the given type in bytes.
    pub fn get_size_of(&self, types: &Types, ty: Type) -> usize {
        match types.get(ty) {
            Some(ty) => match &*ty {
                CompoundType::Array(a) => {
                    self.get_size_of(types, a.inner) * a.num_elements as usize
                }
                CompoundType::Pointer(_) => 8,
                CompoundType::Struct(s) => StructLayout::new(self, types, s).get_size(),
                e => todo!("{:?}", e),
            },
            None => match ty {
                types::VOID => 0,
                types::I1 => 1,
                types::I8 => 1,
                types::I16 => 2,
                types::I32 => 4,
                types::I64 => 8,
                x => todo!("sizeof {:?}", x),
            },
        }
    }

    pub fn get_align_of(&self, types: &Types, ty: Type) -> usize {
        match types.get(ty) {
            Some(ty) => match &*ty {
                CompoundType::Array(a) => self.get_size_of(types, a.inner),
                CompoundType::Pointer(_) => 8,
                CompoundType::Struct(s) => StructLayout::new(self, types, s).get_align(),
                _ => todo!(),
            },
            None => match ty {
                types::VOID => 0,
                types::I1 => 1,
                types::I8 => 1,
                types::I16 => 2,
                types::I32 => 4,
                types::I64 => 8,
                x => todo!("sizeof {:?}", x),
            },
        }
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl From<String> for DataLayout {
    fn from(s: String) -> Self {
        DataLayout(s)
    }
}

impl StructLayout {
    pub fn new(dl: &DataLayout, types: &Types, sty: &StructType) -> Self {
        let mut offset = 0;
        let mut align = 1;
        let mut offsets = vec![];
        for &elem in &sty.elems {
            let elem_align = if sty.is_packed {
                1
            } else {
                dl.get_align_of(types, elem)
            };
            if !is_aligned(elem_align, offset) {
                offset = align_to(offset, elem_align);
            }
            align = align.max(elem_align);
            offsets.push(offset);
            offset += dl.get_size_of(types, elem);
        }
        if !is_aligned(align, offset) {
            offset = align_to(offset, align);
        }
        Self(offsets, offset, align)
    }

    pub fn get_elem_offset(&self, idx: usize) -> Option<usize> {
        self.0.get(idx).copied()
    }

    pub fn get_size(&self) -> usize {
        self.1
    }

    pub fn get_align(&self) -> usize {
        self.2
    }
}

fn is_aligned(align: usize, offset: usize) -> bool {
    offset % align == 0
}

fn align_to(offset: usize, align: usize) -> usize {
    (offset + align - 1) & !(align - 1)
}

#[test]
fn size() {
    use crate::ir::types::*;

    let dl = DataLayout("".to_string());
    let types = Types::new();
    assert_eq!(dl.get_size_of(&types, types::VOID), 0);
    assert_eq!(dl.get_size_of(&types, types::I1), 1);
    assert_eq!(dl.get_size_of(&types, types::I8), 1);
    assert_eq!(dl.get_size_of(&types, types::I16), 2);
    assert_eq!(dl.get_size_of(&types, types::I32), 4);
    assert_eq!(dl.get_size_of(&types, types::I64), 8);

    let i8_ptr = types.base_mut().pointer(types::I8);
    assert_eq!(dl.get_size_of(&types, i8_ptr), 8);

    let i8_arr = types.base_mut().array(ArrayType::new(types::I8, 100));
    assert_eq!(dl.get_size_of(&types, i8_arr), 100);

    let i32_arr = types.base_mut().array(ArrayType::new(types::I32, 100));
    assert_eq!(dl.get_size_of(&types, i32_arr), 400);

    let strukt0 = types
        .base_mut()
        .anonymous_struct(vec![types::I32, types::I32], false);
    assert_eq!(dl.get_size_of(&types, strukt0), 8);
    let strukt1 = types
        .base_mut()
        .anonymous_struct(vec![types::I64, types::I64, types::I64], false);
    assert_eq!(dl.get_size_of(&types, strukt1), 24);
    let strukt2 = types
        .base_mut()
        .anonymous_struct(vec![types::I64, types::I64, types::I32], false);
    assert_eq!(dl.get_size_of(&types, strukt2), 24);
    let strukt3 = types
        .base_mut()
        .anonymous_struct(vec![types::I8, strukt2, types::I32], false);
    assert_eq!(dl.get_size_of(&types, strukt3), 40);
}

#[test]
fn align() {
    use crate::ir::types::*;

    let dl = DataLayout("".to_string());
    let types = Types::new();
    assert_eq!(dl.get_align_of(&types, types::VOID), 0);
    assert_eq!(dl.get_align_of(&types, types::I1), 1);
    assert_eq!(dl.get_align_of(&types, types::I8), 1);
    assert_eq!(dl.get_align_of(&types, types::I16), 2);
    assert_eq!(dl.get_align_of(&types, types::I32), 4);
    assert_eq!(dl.get_align_of(&types, types::I64), 8);

    let i8_ptr = types.base_mut().pointer(types::I8);
    assert_eq!(dl.get_align_of(&types, i8_ptr), 8);

    let i8_arr = types.base_mut().array(ArrayType::new(types::I8, 100));
    assert_eq!(dl.get_align_of(&types, i8_arr), 1);

    let i32_arr = types.base_mut().array(ArrayType::new(types::I32, 100));
    assert_eq!(dl.get_align_of(&types, i32_arr), 4);

    let strukt0 = types
        .base_mut()
        .anonymous_struct(vec![types::I32, types::I32], false);
    assert_eq!(dl.get_align_of(&types, strukt0), 4);
    let strukt1 = types
        .base_mut()
        .anonymous_struct(vec![types::I64, types::I64, types::I64], false);
    assert_eq!(dl.get_align_of(&types, strukt1), 8);
    let strukt2 = types
        .base_mut()
        .anonymous_struct(vec![types::I64, types::I64, types::I32], false);
    assert_eq!(dl.get_align_of(&types, strukt2), 8);
    let strukt3 = types
        .base_mut()
        .anonymous_struct(vec![types::I8, strukt2, types::I32], false);
    assert_eq!(dl.get_align_of(&types, strukt3), 8);
}
