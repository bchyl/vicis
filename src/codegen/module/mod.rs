use super::{function::Function, isa::TargetIsa};
use crate::ir::{
    module::{
        attributes::Attribute, global_variable::GlobalVariable, name::Name, Target as ModuleTarget,
    },
    types::Types,
};
use id_arena::Arena;
use rustc_hash::FxHashMap;

pub struct Module<T: TargetIsa> {
    pub name: String,
    pub source_filename: String,
    pub target: ModuleTarget, // TODO
    pub functions: Arena<Function<T>>,
    pub attributes: FxHashMap<u32, Vec<Attribute>>,
    pub global_variables: FxHashMap<Name, GlobalVariable>,
    pub types: Types,
    // TODO: Metadata
    pub arch: T, // TODO: Should be named 'target'
}
