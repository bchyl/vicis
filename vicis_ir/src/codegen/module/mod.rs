use super::{function::Function, isa::TargetIsa};
use crate::ir::{
    module::{
        attributes::Attribute, global_variable::GlobalVariable, name::Name, Target as ModuleTarget,
    },
    types::Types,
};
use id_arena::Arena;
use rustc_hash::FxHashMap;
use std::fmt;

pub struct Module<T: TargetIsa> {
    pub name: String,
    pub source_filename: String,
    pub target: ModuleTarget, // TODO
    pub functions: Arena<Function<T>>,
    pub attributes: FxHashMap<u32, Vec<Attribute>>,
    pub global_variables: FxHashMap<Name, GlobalVariable>,
    pub types: Types,
    // TODO: Metadata
    pub isa: T,
}

impl<T: TargetIsa> fmt::Debug for Module<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "source_filename = \"{}\"", self.source_filename)?;
        writeln!(f, "target datalayout = \"{}\"", self.target.datalayout())?;
        writeln!(f, "target triple = \"{}\"", self.target.triple())?;
        writeln!(f)?;
        for gv in self.global_variables.values() {
            writeln!(f, "{}", gv.to_string(&self.types))?;
        }
        writeln!(f)?;
        for (_, func) in &self.functions {
            writeln!(f, "{:?}", func)?;
        }
        for (id, attrs) in &self.attributes {
            write!(f, "attributes #{} = {{ ", id)?;
            for attr in attrs {
                write!(f, "{:?} ", attr)?;
            }
            writeln!(f, "}}")?
        }
        Ok(())
    }
}
