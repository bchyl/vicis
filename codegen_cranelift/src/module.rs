use crate::{
    function::{compile_function, declare_and_define_function},
    LowerCtx,
};
use cranelift_codegen::Context;
use cranelift_module::Module;
use vicis_core::ir::module::Module as LlvmModule;

pub fn compile_module<M: Module>(clif_mod: &mut M, clif_ctx: &mut Context, llvm_mod: &LlvmModule) {
    let mut lower_ctx = LowerCtx::new(llvm_mod, clif_mod);

    let mut funcs = vec![];
    let mut protos = vec![];

    for func in llvm_mod.functions() {
        if func.1.is_prototype() {
            protos.push(func);
        } else {
            funcs.push(func);
        }
    }

    // Declare prototypes first.
    for (func_id, _) in protos {
        compile_function(&mut lower_ctx, clif_ctx, func_id);
        lower_ctx.clif_mod.clear_context(clif_ctx);
    }

    // Define functions.
    for (func_id, func) in funcs {
        compile_function(&mut lower_ctx, clif_ctx, func_id);
        declare_and_define_function(lower_ctx.clif_mod, clif_ctx, func.name().as_str());
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test1() {
        compile(
            r#"
        declare dso_local i32 @putchar(i8 signext)
        define dso_local i32 @main() {
          call i32 @putchar(i8 signext 65)
          ret i32 0
        }"#,
        );
    }

    fn compile(source: &str) {
        use cranelift::prelude::Configurable;
        use cranelift_codegen::{isa, settings};
        use cranelift_object::{ObjectBuilder, ObjectModule};
        use vicis_core::ir::module::Module;

        let mut flag_builder = settings::builder();
        flag_builder.enable("is_pic").unwrap();
        #[cfg(all(target_os = "linux", target_arch = "aarch64"))]
        let isa_builder = isa::lookup_by_name("aarch64-unknown-unknown-elf").unwrap();
        #[cfg(all(target_os = "linux", target_arch = "x86_64"))]
        let isa_builder = isa::lookup_by_name("x86_64-unknown-unknown-elf").unwrap();
        #[cfg(all(target_os = "macos", target_arch = "aarch64"))]
        let isa_builder = isa::lookup_by_name("aarch64-apple-darwin").unwrap();
        let isa = isa_builder.finish(settings::Flags::new(flag_builder));

        let builder = ObjectBuilder::new(
            isa,
            "".to_owned(), // TODO: This will be embedded in the object file.
            cranelift_module::default_libcall_names(),
        )
        .unwrap();
        let mut clif_mod = ObjectModule::new(builder);
        let mut clif_ctx = clif_mod.make_context();

        let module = Module::try_from(source).unwrap();
        compile_module(&mut clif_mod, &mut clif_ctx, &module);

        clif_mod.finish();
    }
}
