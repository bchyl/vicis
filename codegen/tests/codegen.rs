use std::fs;
use vicis_codegen::{isa::x86_64::X86_64, lower::compile_module};
use vicis_core::ir::module::Module;

macro_rules! test {
    ($isa:ident, $testname:ident, $name:expr) => {
        #[test]
        fn $testname() {
            let parent = "./tests/codegen/";
            let input = format!("{}{}.ll", parent, $name);
            let input_body = &fs::read_to_string(input).unwrap();
            let module = Module::try_from(input_body.as_str()).unwrap();
            let isa = $isa::default();
            let mach_module = compile_module(&isa, &module).unwrap();
            insta::assert_display_snapshot!(mach_module.display_asm());
        }
    };
}

#[cfg(test)]
mod x86_64 {
    use super::*;

    test!(X86_64, test_add, "ary1");
    test!(X86_64, test_ary2, "ary2");
    test!(X86_64, test_ary3, "ary3");
    test!(X86_64, test_ary4, "ary4");
    test!(X86_64, test_ary5, "ary5");
    test!(X86_64, test_br, "br");
    test!(X86_64, test_call1, "call1");
    test!(X86_64, test_call2, "call2");
    test!(X86_64, test_condbr, "condbr");
    test!(X86_64, test_fibo, "fibo");
    test!(X86_64, test_load_add, "load_add");
    test!(X86_64, test_phi, "phi");
    test!(X86_64, test_phi2, "phi2");
    test!(X86_64, test_puts, "puts");
    test!(X86_64, test_sum, "sum");
    test!(X86_64, test_hello, "hello");
    test!(X86_64, test_addr, "addr");
    test!(X86_64, test_i8, "i8");
    test!(X86_64, test_i8_load_store, "i8_load_store");
    test!(X86_64, test_global, "global");
    test!(X86_64, test_spill, "spill");
}
