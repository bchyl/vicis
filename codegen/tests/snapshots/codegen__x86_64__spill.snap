---
source: codegen/tests/codegen.rs
assertion_line: 44
expression: mach_module.display_asm()
---
  .text
  .intel_syntax noprefix
  .text
  .globl main
main:
.LBL0_0:
  push rbp
  mov rbp, rsp
  sub rsp, 64
  mov dword ptr [rbp-4], 1
  mov dword ptr [rbp-8], 2
  mov dword ptr [rbp-12], 3
  mov dword ptr [rbp-16], 4
  mov dword ptr [rbp-20], 5
  mov dword ptr [rbp-24], 6
  mov dword ptr [rbp-28], 7
  mov dword ptr [rbp-32], 8
  mov dword ptr [rbp-36], 9
  mov dword ptr [rbp-40], 10
  mov eax, dword ptr [rbp-4]
  mov ecx, dword ptr [rbp-8]
  mov edx, dword ptr [rbp-12]
  mov edi, dword ptr [rbp-16]
  mov esi, dword ptr [rbp-20]
  mov dword ptr [rbp-44], esi
  mov esi, dword ptr [rbp-24]
  mov dword ptr [rbp-48], esi
  mov esi, dword ptr [rbp-28]
  mov dword ptr [rbp-52], esi
  mov esi, dword ptr [rbp-32]
  mov dword ptr [rbp-56], esi
  mov esi, dword ptr [rbp-36]
  mov dword ptr [rbp-60], esi
  mov esi, dword ptr [rbp-40]
  mov dword ptr [rbp-64], esi
  add eax, ecx
  add eax, edx
  add eax, edi
  mov esi, dword ptr [rbp-44]
  add eax, esi
  mov esi, dword ptr [rbp-48]
  add eax, esi
  mov esi, dword ptr [rbp-52]
  add eax, esi
  mov esi, dword ptr [rbp-56]
  add eax, esi
  mov esi, dword ptr [rbp-60]
  add eax, esi
  mov esi, dword ptr [rbp-64]
  add eax, esi
  add rsp, 64
  pop rbp
  ret 

